### PR TITLE
[ML] Multiphase progress reporting for data frame analyses

### PR DESCRIPTION
**Pull request by tveasey**
This implements phased progress reporting for data frame analytics. I've extended `CDataFrameAnalysisInstrumentation` to support starting progress monitoring for named tasks. In order to ensure that we always send a phase complete result I also moved the progress writing out of `CDataFrameAnalyzer` into `CDataFrameAnalysisInstrumentation`. (This seems more natural anyway.)

### CODE REVIEW

File: lib/maths/CBoostedTreeImpl.cc [hunk 1]
```diff
@@ -1463,8 +1494,6 @@ bool CBoostedTreeImpl::acceptRestoreTraverser(core::CStateRestoreTraverser& trav
                                              m_MaximumNumberTrees, traverser))
         RESTORE(TRAINING_ROW_MASKS_TAG,
                 core::CPersistUtils::restore(TRAINING_ROW_MASKS_TAG, m_TrainingRowMasks, traverser))
-        RESTORE(TRAINING_PROGRESS_TAG,
```
Comment (dimitris-athanasiou) [inline lib/maths/CBoostedTreeImpl.cc:1466]: Will we safely ignore this tag in model states from previous versions?

File: lib/maths/COutliers.cc [hunk 2]
```diff
@@ -33,7 +33,7 @@ using namespace outliers_detail;
 
 namespace {
 
-const std::string COMPUTE_OUTLIER_SCORES{"compute_outlier_scores"};
+const std::string COMPUTE_OUTLIERS{"computing_outliers"};
```
Comment (dimitris-athanasiou) [inline lib/maths/COutliers.cc:36]: if we decide to match labels with content we should do it here too

File: lib/maths/CBoostedTreeImpl.cc (hunk 1 above)
Comment (tveasey) [reply lib/maths/CBoostedTreeImpl.cc:1466]: Yes if a tag turns up we don't know about we just ignore in the restore code.

File: lib/maths/CBoostedTreeFactory.cc [hunk 3]
```diff
@@ -1243,5 +1237,10 @@ std::size_t CBoostedTreeFactory::mainLoopMaximumNumberTrees(double eta) const {
 
 void CBoostedTreeFactory::noopRecordTrainingState(CBoostedTree::TPersistFunc) {
 }
+
+const std::string CBoostedTreeFactory::FEATURE_SELECTION{"feature_selection"};
+const std::string CBoostedTreeFactory::COARSE_PARAMETER_SEARCH{"coarse_parameter_search"};
+const std::string CBoostedTreeFactory::FINE_TUNE_PARAMETERS{"fine_tuning_parameters"};
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CBoostedTreeFactory.cc]: See [3d0b42e](https://github.com/elastic/ml-cpp/pull/1179/commits/3d0b42e404747cdfc195c609f369118de806fafc).

File: lib/maths/CBoostedTreeImpl.cc (hunk 1 above)
Comment (tveasey) [reply lib/maths/CBoostedTreeImpl.cc:1466]: See [3d0b42e](https://github.com/elastic/ml-cpp/pull/1179/commits/3d0b42e404747cdfc195c609f369118de806fafc).

File: lib/maths/COutliers.cc (hunk 2 above)
Comment (tveasey) [reply lib/maths/COutliers.cc]: See [3d0b42e](https://github.com/elastic/ml-cpp/pull/1179/commits/3d0b42e404747cdfc195c609f369118de806fafc).

File: lib/maths/CBoostedTreeImpl.cc [hunk 4]
```diff
@@ -1272,6 +1279,32 @@ void CBoostedTreeImpl::recordHyperparameters() {
             m_Regularization.leafWeightPenaltyMultiplier()};
 }
 
+void CBoostedTreeImpl::startProgressMonitoringFineTuneHyperparameters() {
+
+    // The this costs "number folds" * "maximum number trees per forest" units
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CBoostedTreeImpl.cc]: See [f409b9a](https://github.com/elastic/ml-cpp/pull/1179/commits/f409b9add4801ac9bb3a06611bdb23731a8165f0).

**Comment by tveasey** [comment]
Thanks @dimitris-athanasiou! I've addressed all your comments. Could you take another look.

**Comment by tveasey** [comment]
retest

**Comment by tveasey** [comment]
retest

**Pull request merged**